### ISSUE TITLE
Livelocks in Microbench.c

### ISSUE THREAD

**Issue by rrnewton** [iu-parfunc/concurrent_cilk#6]
A basic way to (probabilistic) increase the # of threads that block in this benchmark is to flip the writers to be spawned after the readers:

```
cilk_spawn readers(all_ivars, num_fibers);
cilk_spawn writer(all_ivars, num_fibers);
```

An even more aggressive (and predictable) way is to either spin or block on on IVar after launching readers but before launching writers. 

All three of these methods for increasing blocking (flip, spin, block) led to a livelocks when we tested it today on the `concurrent_cilk` branch (68e58a44).  With `CILK_NWORKERS=4`, the app would spin at 400% CPU but with flat memory usage.

**Comment by czakian** [comment]
I tried to reproduce this, but it was far from consistent. 

gdb says that there runtime state believes there is blocked ivar which has not yet been written:

#3  __cilkrts_scheduler (w=0x6384c0) at /home/czakian/Research/concurrent_cilk/runtime/scheduler.c:1596
(gdb) p w
$3 = (__cilkrts_worker *) 0x6384c0
(gdb) p \* 
$4 = {tail = 0x638b00, head = 0x638b00, exc = 0x638b00, protected_tail = 0x63ab00, ltq_limit = 0x63ab00, self = 3, 
  g = 0x7ffff79bb628 <(anonymous namespace)::global_state_singleton>, l = 0x6385c0, reducer_map = 0x0, current_stack_frame = 0x0, reserved = 0x0, sysdep = 0x0, pedigree = {
    rank = 0, parent = 0x0}, paused_ctx = 0x0, team_leader = 0x6055c0, referencelist = 0x65abe0, readylist = 0x64ab10, fibers = 0x642b00, blocked = 0, worker_depth = 1, 
  ref_count = 0x66ad00}
(gdb) p w
$5 = (__cilkrts_worker *) 0x6384c0
(gdb) p w->ref_count 
$6 = (int \* volatile) 0x66ad00
(gdb) p *
$7 = 1

**Comment by czakian** [comment]
And here's the blocked worker: 

gdb) p w->fibers  
$8 = (__cilkrts_worker \* volatile *) 0x642b00
(gdb) p *
$9 = (__cilkrts_worker \* volatile) 0x6055c0
(gdb) p *
$10 = {tail = 0x615450, head = 0x615450, exc = 0x615450, protected_tail = 0x617440, ltq_limit = 0x617440, self = 3, 
  g = 0x7ffff79bb628 <(anonymous namespace)::global_state_singleton>, l = 0x614f00, reducer_map = 0x610080, current_stack_frame = 0x7fffffffd758, reserved = 0x0, 
  sysdep = 0x0, pedigree = {rank = 0, parent = 0x7fffffffd7a8}, paused_ctx = 0x7fffffffd650, team_leader = 0x6055c0, referencelist = 0x65abe0, readylist = 0x64ab10, 
  fibers = 0x642b00, blocked = 1, worker_depth = 0, ref_count = 0x66ad00}

**Comment by czakian** [comment]
By manually moving the worker to the ready dequeue, I was able to unbind the process and complete the execution without incident. My suspicion is either a data race in the write while the reader is transitioning state (thus the reader misses out on being enqueued into the ready list) or that the cilk_sync is have a consequence we didn't expect (less likely).

**Comment by czakian** [comment]
I checked my original theory today and realized that the conditions for my hypothesis to happen don't exist because there is only a single reader of every ivar. I took another look at the microbench code and I believe this issue was user error while writing the microbench.c code. I have removed an erroneous ivar_clear() call in the writer which clobbered the state stored in the ivar by a proceeding pause. I have commited the fix as of commit: https://github.com/iu-parfunc/cilk_tests/commit/fc51ddee60f525212cce8493cbc419835e49b42e
